oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/NodeDocumentSweeper2.java

```diff
@@ -0,0 +1,357 @@
+            throws DocumentStoreException {
+        head = headRevision.getRevision(clusterId);
+        totalCount = 0;
+        lastCount = 0;
+        startOfScan = context.getClock().getTime();
+        lastLog = startOfScan;
+
+        if (head == null) {
+            LOG.warn("Head revision does not have an entry for " +
+                            "clusterId {}. Sweeping of documents is skipped.",
+                    clusterId);
+            return;
+        }
+
+        Iterable<Map.Entry<Path, UpdateOp>> ops = sweepOperations(documents);
+        for (List<Map.Entry<Path, UpdateOp>> batch : partition(ops, INVALIDATE_BATCH_SIZE)) {
+            Map<Path, UpdateOp> updates = newHashMap();
+            for (Map.Entry<Path, UpdateOp> entry : batch) {
+                updates.put(entry.getKey(), entry.getValue());
+            }
+            listener.sweepUpdate(updates);
+        }
+        LOG.debug("Document sweep finished");
+    }
+
+    private Iterable<Map.Entry<Path, UpdateOp>> sweepOperations(
+            final Iterable<NodeDocument> docs) {
+        return filter(transform(docs,
+                new Function<NodeDocument, Map.Entry<Path, UpdateOp>>() {
+
+            int yieldCnt = 0;
+            long lastYield = System.currentTimeMillis();
+
+            @Override
+            public Map.Entry<Path, UpdateOp> apply(NodeDocument doc) {
+                if (++yieldCnt >= YIELD_SIZE) {
+                    try {
+                        Thread.sleep(Math.max(1, System.currentTimeMillis() - lastYield));
+                    } catch (InterruptedException e) {
+                        // ignore
+                    }
+                    lastYield = System.currentTimeMillis();
+                    yieldCnt = 0;
+                }
+                return immutableEntry(doc.getPath(), sweepOne(doc));
+            }
+        }), new Predicate<Map.Entry<Path, UpdateOp>>() {
+            @Override
+            public boolean apply(Map.Entry<Path, UpdateOp> input) {
+                return input.getValue() != null;
+            }
+        });
+    }
+
+    private UpdateOp sweepOne(NodeDocument doc) throws DocumentStoreException {
+        UpdateOp op = createUpdateOp(doc);
+        // go through PROPERTY_OR_DELETED_OR_COMMITROOT, whereas :
+        // - PROPERTY : for content changes
+        // - DELETED : for new node (this)
+        // - COMMITROOT : for new child (parent)
+        // - REVISIONS : for commit roots (root for branch commits)
+        for (String property : filter(doc.keySet(), COMMITROOT_OR_REVISIONS)) {
+            Map<Revision, String> valueMap = doc.getLocalMap(property);
+            for (Map.Entry<Revision, String> entry : valueMap.entrySet()) {
+                Revision rev = entry.getKey();
+
+                // consider any clusterId for sweep2
+
+                Revision cRev = getCommitRevision(doc, rev);
+                if (cRev == null) {
+                    uncommitted(doc, property, rev, op);
```

Comment:
This will remove valid committed changes!

NodeDocumentSweeper2 runs on a CommitValueResolver without sweep revisions. Starting with 1.8 and introduction of the sweep revision, the revision garbage collector is allowed to remove split documents with commit information older than the sweep revision. The changes referring to the removed commit information must still be considered committed because of to the sweep revision.

Reply:
Very good and important find. This class was not mean to replicate actual sweep[1] behaviour but only apply missing `_bc` - definitely it should not uncommit anything. Will add a test-case to verify this would have uncommited and fix accordingly.

Reply:
Removed the uncommit-cleanup logic and added a test case to cover this.